### DISCUSSION TITLE
Font Face Selection for Hangul Normalized in NFD

### DISCUSSION THREAD

**Discussion by Pusnow** [ghostty-org/ghostty#4163] [Q&A]
## Background

To represent Hangul (Korean letters) in unicode, the two normalization forms (NFC and NFD) are widely used.
* Using NFC (widely used in Windows/Linux), every syllable is represented in a single unicode point.
* Using NFD (widely used in Mac), a syllable is represented by combining two or three unicode points.

For example, Korean syllable '각' can be represented by either

* NFC: U+AC01
* NFD: U+1100,U+1161,U+11A8

## Problem

Ghostty seems to select font face by using code points before normalization and doesn't enforce font face even if I've configured `font-codepoint-map = "U+1100-U+11FF=D2Coding"`. (D2Coding is a widely used monospaced font for Korean and Hangul.)
D2Coding doesn't have code points U+1100-U+11FF.

I want to use the D2Coding font for hangul text normalized in NFD.
I'm not sure where  (ligature in font?, OS?, app) to implement this, but there is a reddit comment that says OS (or some font redering layer in Ghostty?) should handle this:
https://www.reddit.com/r/FontForge/comments/ongyqo/comment/h5s8g8w/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

Maybe, font rendering considering Unicode equivalence would solve this problem.
https://en.wikipedia.org/wiki/Unicode_equivalence

### Config

```
font-family = D2Coding
font-size = 24
font-codepoint-map = "U+1100-U+11FF=D2Coding"
selection-invert-fg-bg = true
```

### ghosttyty +show-face result

U+1100 « ᄀ » found in face “PCMyungjo”.
U+1161 «  » found in face “PCMyungjo”.
U+11A8 «  » found in face “PCMyungjo”.

### Font rendering in Ghostty

The first line is normalized in NFC (rendered in D2Coding (sans-serif)) and the third line is normalized in NFD (rendered in PCMyungjo (serif)).

<img width="249" alt="스크린샷 2024-12-31 오후 3 21 20" src="https://github.com/user-attachments/assets/0433dd29-4b0c-42fc-b442-1778c7a2a4a6" />

### Font rendering in Mac's Terminal

<img width="241" alt="스크린샷 2024-12-31 오후 3 21 36" src="https://github.com/user-attachments/assets/0a4581fd-7d3a-4c3b-bd83-596491c0067f" />

### File used for test

[test.txt](https://github.com/user-attachments/files/18279666/test.txt)

**Comment by kingori** [comment]
+1 to this issue.  When I create a directory in OSX's finder, it is showing dirrently in ls output on Ghostty.

<img width="781" height="361" alt="image" src="https://github.com/user-attachments/assets/cf10cc16-170e-4bab-854e-3b26f0422ad5" />

In above screenshot, I created "테스트" (means test in Korean) on Finder and created "테스트2" on terminal using mkdir command. As Finder deals utf-8 diffently, "테스트" looks differently.

I tried to set `font-codepoint-map` , but it does not work.

```
# 1. 한글 자모 (NFD - macOS가 한글을 저장/출력할 때 쓰는 쪼개진 자모 영역)
font-codepoint-map = U+1100-U+11FF="D2CodingLigature Nerd Font Mono"

# 2. 한글 호환 자모 (단독으로 쓰이는 ㄱ, ㄴ, ㄷ 영역)
font-codepoint-map = U+3130-U+318F="D2CodingLigature Nerd Font Mono"

# 3. 한글 음절 (NFC - 완성형 '가, 각, 간' 영역)
font-codepoint-map = U+AC00-U+D7A3="D2CodingLigature Nerd Font Mono"

# 4. 한글 자모 확장 (드물게 쓰이는 옛한글 등 확장 영역)
font-codepoint-map = U+A960-U+A97F="D2CodingLigature Nerd Font Mono"
font-codepoint-map = U+D7B0-U+D7FF="D2CodingLigature Nerd Font Mono"
``` 

```
> ghostty +show-face --string="테스 ""
U+1110 « ᄐ » found in face “PCMyungjo”.
U+1166 «  » found in face “PCMyungjo”.
U+1109 « ᄉ » found in face “PCMyungjo”.
U+1173 «  » found in face “PCMyungjo”.
U+1110 « ᄐ » found in face “PCMyungjo”.
U+1173 «  » found in face “PCMyungjo”.
```

**Reply by exploitJ** [reply]
use Noto Sans Mono CJK KR font for overriding those codepoints.
